File: java/hash/Hash.java
import java.util.List;

public class Hash {
    public  int tableSize;
    public  int overflowSize;
    private int realsize;
    private int posOverflow; // control insertions on overflow area
    public Object htable[]; // make easy to substitute the type of the table

    public Hash() {
        this(10, 10);
    }

    public Hash(int tableSize, int overflowSize) {
        this.tableSize = tableSize;
        this.overflowSize = overflowSize;
        this.realsize = tableSize + overflowSize;
        this.posOverflow = 0;
        this.htable = new Object[realsize];
        for (int i = 0; i < realsize; i++) 
            htable[i] = null;
    }

    private int hash(Object item) {
        // change the key with whatever you want
        return ((int)item % tableSize);
    }

    // return the position where item has been inserted
    public int insert(Object item) {
        int pInsertion = hash(item); 
        int posReturn = -1; 
        // if the position is free
        if (htable[pInsertion] == null) {
            htable[pInsertion] = item;
            posReturn = pInsertion;
        } else {
            if (posOverflow < (realsize - tableSize)) {
                htable[tableSize + posOverflow] = item;
                posReturn = tableSize + posOverflow;
                posOverflow++;
            } else {
                System.out.print("Área de reserva cheia! Falha ao inserir "+(int)item);
            }
        }
        return posReturn;
    }

	public void insertList(List<Integer> lista) {
		for (int item : lista) 
			insert((Object) item);
	}

    // return the position of the removed element
    // -1 if not found
    public int remove(Object toBeRemoved) {
        int pos = -1;
        int pRemotion = hash(toBeRemoved);
        // found in the main area
        if ((int)htable[pRemotion] == (int)toBeRemoved) {
            pos = pRemotion;
            boolean hasSubstitute = false;
            // change the item in the main area
            // with the item in the overflow area
            // with the same hash 
            for (int i = 0; i < posOverflow && !hasSubstitute; i++) {
                if (hash(htable[i + tableSize]) == pRemotion) {
                    hasSubstitute = true;
                    htable[pRemotion] = htable[i + tableSize];
                    if (posOverflow > 0) 
                        posOverflow--;
                    htable[i + tableSize] = htable[tableSize + posOverflow];
                    htable[tableSize + posOverflow] = null;
                }
            }
        } else {
            // not found in main area -> search in the overflow area
            for (int i = 0; i < posOverflow; i++) {
                if ((int)htable[i + tableSize] == (int)toBeRemoved) {
                    pos = i + tableSize;
                    if (posOverflow > 0)
                        posOverflow--;
                    htable[i + tableSize] = htable[posOverflow];
                    htable[posOverflow] = null;
                    i = posOverflow;
                }
            }
        }
        return pos;
    }

    // return the removed item
    // null if not found
    public Object remove(int key) {
        if (key < 0 || key >= realsize) {
            System.out.println("Chave de remocao invalida.");
            return null;
        }

        Object removed = null;
        if (htable[key] != null) {
            removed = htable[key];
            boolean hasSubstitute = false;
            // again, put in the main area
            // the item with the same hash coming
            // from the overflow area
            for (int i = 0; i < posOverflow && !hasSubstitute; i++) {
                if (hash(htable[i + tableSize]) == key) {
                    hasSubstitute = true;
                    htable[key] = htable[i + tableSize];
                    if (posOverflow > 0) 
                        posOverflow--;
                    htable[i + tableSize] = htable[posOverflow];
                    htable[posOverflow] = null;
                }
            }
        }
        return removed;
    }

    public void showItens() {
        System.out.println("\nÁrea principal: [");
        for (int i = 0; i < tableSize; i++) {
            System.out.print("\t");
            if (htable[i] != null)
                System.out.println((int)htable[i]);
            else
                System.out.println("null");
        }
        System.out.println("]\n");
        System.out.println("Área de reserva: [");
        for (int i = 0; i < posOverflow; i++) {
            System.out.print("\t"+(int)htable[tableSize + i] + "\n");
        }
        System.out.println("]\n");
    }

    public boolean search(Object obj) {
        boolean retorno = false;
        if (obj == null) {
            System.out.println("Item não encontrado.");
        } else {
            int hashJ = hash(obj);
            if ((int)htable[hashJ] == (int)obj) {
                retorno = true;
            } else {
                // searching on overflow 
                for (int i = 0; i < posOverflow && !retorno; i++) {
                    if ((int)htable[i + tableSize] == (int)obj) {
                        retorno = true;
					}
                }
            }
        }
        return retorno;
    }

    // return the position of the item
    // -1 if not found
    public int searchPos(Object obj) {
        int pos = -1;
        if (obj == null) {
            System.out.println("Item não encontrado.");
        } else {
            int hashJ = hash(obj);
            if ((int)htable[hashJ] == (int)obj) {
                pos = hashJ;
            } else {
                // searching in overflow area
                for (int i = 0; i < posOverflow && (pos == -1); i++) {
                    if ((int)htable[i + tableSize] == (int)obj) {
                        pos = i + tableSize;
                    }
                }
            }
        }
        return pos;
    }
}

File: c/hash.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

typedef struct Hash {
	int tableSize;
	int overflowSize;
	int realSize;
	int posOverflow; // control insertions in overflow area
	int ** thash;
} Hash;

Hash * newHash(int tamTable, int tamOverflow) {
	Hash * th = (Hash *) malloc(sizeof(Hash));

	th->posOverflow = 0;
	th->tableSize = tamTable;
	th->overflowSize = tamOverflow;
	th->realSize = tamTable + tamOverflow;

    th->thash = (int **) malloc(sizeof(int *) * th->realSize);

    for (int i = 0; i < th->realSize; i++) {
        th->thash[i] = NULL;
    }

    return th;
}

int hash(Hash * th, int key) {
    return key % th->tableSize;
}

// return the position where the item
// has been inserted
// return -1 if could not insert
int inserir(Hash * th, int item) {
    int posInsercao = -1;
    int pos = hash(th, item);
    int *p = (int *) malloc(sizeof(int));
    *p = item;
    if (th->thash[pos] == NULL) {
        th->thash[pos] = p;
        posInsercao = pos;
    } else {
		// check if there is space to insert something
		// in the overflow area
        if (th->posOverflow < (th->realSize - th->tableSize)) {
            int posReserva = th->tableSize + th->posOverflow++;
            th->thash[posReserva] = p; 
            posInsercao = posReserva;
        } else {
            printf("Area de reserva cheia!\n");
            free(p);
        }
    }
    p = NULL;
    return posInsercao;
}

// remove item using its position
int removerPos(Hash * th, int key) {
    int removed = 0x7fffffff;

    // is key valid?
    if (key < 0 || key >= th->realSize) {
        printf("Chave de procura invalida. Chave: %d\n", key);
    } else {
        // if there is something in the position
        if (th->thash[key] != NULL) {
            removed = *th->thash[key];
            bool hasSubstitute = false;
			// search for a substitute that has
			// the same hash of the item
            for (int i = 0; i < th->posOverflow && !hasSubstitute; i++) {
                if (hash(th, *th->thash[i + th->tableSize]) == key) {
                    hasSubstitute = true;
                    *(th->thash[key]) = *(th->thash[i + th->tableSize]);
                    if (th->posOverflow > 0)
                        th->posOverflow--;
                    *(th->thash[i + th->tableSize]) = *(th->thash[th->tableSize + th->posOverflow]);
                    free(th->thash[th->tableSize + th->posOverflow]);
                    th->thash[th->tableSize + th->posOverflow] = NULL;
                }
            }
			// if there is no substitute,
			// just remove the item
            if (!hasSubstitute) {
                free(th->thash[key]);
                th->thash[key] = NULL;
            }
        }
    }

    return removed;
}

// remove item -> return its position 
int removerItem(Hash * th, int item) {
    int posRemovido = -1;
    int posItem = hash(th, item);
    if (*(th->thash[posItem]) == item) {
        posRemovido = posItem;
        bool hasSubstitute = false;
        for (int i = 0; i < th->posOverflow && !hasSubstitute; i++) {
            if (hash(th, *(th->thash[i + th->tableSize])) == posItem) {
                hasSubstitute = true;
                *(th->thash[posItem]) = *(th->thash[i + th->tableSize]);
                if (th->posOverflow > 0)
                    th->posOverflow--;
                *(th->thash[i + th->tableSize]) = *(th->thash[th->tableSize + th->posOverflow]);
                free(th->thash[th->tableSize + th->posOverflow]);
                th->thash[th->tableSize + th->posOverflow] = NULL;
            }
        }
        if (!hasSubstitute) {
            free(th->thash[posItem]);
            th->thash[posItem] = NULL;
        }
    } else {
        // searching in the overflow area
        for (int i = 0; i < th->posOverflow; i++) {
            if (*(th->thash[i + th->tableSize]) == item) {
                posRemovido = i + th->tableSize;
                // if th->posOverflow > 0, the position of
                // the last item in overflow area
                // is th->tableSize + th->posOverflow - 1
                if (th->posOverflow > 0)
                    th->posOverflow--;
                *(th->thash[i + th->tableSize]) = *(th->thash[th->tableSize + th->posOverflow]);
                free(th->thash[th->tableSize + th->posOverflow]);
                th->thash[th->tableSize + th->posOverflow] = NULL;
                i = th->posOverflow;
            }
        }
    }
    return posRemovido;
}

void mostrar(Hash * th) {
    printf("\nÁrea principal: [ ");
    for (int i = 0; i < th->tableSize; i++) 
        printf("%d ", (th->thash[i] == NULL) ? -1 : *(th->thash[i]));
    printf("]\n");
    
    printf("Área de reserva: [ ");
    for (int i = 0; i < th->overflowSize; i++) 
        printf("%d ", (th->thash[i + th->tableSize] == NULL) ? -1 : *(th->thash[i + th->tableSize]));
    printf("]\n\n");
}

int main() {
	Hash * th = newHash(7, 5);
    inserir(th, 10);
    inserir(th, 30);
    mostrar(th);
    inserir(th, 20);
    inserir(th, 21);
    inserir(th, 14);
    inserir(th, 45);
    inserir(th, 35);
    inserir(th, 4);
    inserir(th, 12);
    inserir(th, 1);
    mostrar(th);
    printf("Posicao removido: %d\n", removerItem(th, 21));
    printf("apos remocao do 21: \n");
    mostrar(th);
    printf("Posicao removido: %d\n", removerItem(th, 45));
	mostrar(th);
    printf("Item removido: %d\n", removerPos(th, 30));
    mostrar(th);
}
